### PR TITLE
Move EIP-2315 to Review

### PR DESCRIPTION
**Pull request by gcolvin**
When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

### CODE REVIEW

File: EIPS/eip-2315.md
Comment (lightclient) [inline EIPS/eip-2315.md:4]: I think the "this proposal" can be dropped as it can be inferred by the context of the description.

**Review by lightclient — CHANGES_REQUESTED** [review]
Please move the `Test Cases` section above `Security Considerations`

**Review by MicahZoltu — CHANGES_REQUESTED** [review]
This currently depends on 3540 (Review), 3670 (Review), 3779 (Draft), and 4200 (Draft).  Any dependencies need to be at least as far along in the EIP process as the dependent, which means this can't move to Review until 3779 and 4200 are in Review.  I didn't dig deep into the technical details for those dependencies, but if they aren't strictly required then an appropriate course of action may be to just remove the dependency.  If you decide to go that route, you'll need to remove the external link to EIP-3779 and EIP-4200 in the Specification and Backward Compatibility sections.

----

External links should be removed before merging into Review.  Specifically this line:
> The [concept](https://people.cs.clemson.edu/~mark/subroutines.html) goes back to [Turing's Automatic Computing Engine of 1946](http://www.alanturing.net/turing_archive/archive/p/p01/P01-001.html):

I recommend just keeping the text and the attached quote, without the links.  If someone wants to look up the source of the quote they can just do an internet search for it, though that shouldn't matter because the argument should stand on its own without needing to appeal to an authority (especially one from 80 years ago).

----

I personally recommend removing the entire History section.  I generally let people include just about whatever they want in the Motivation (aside from external links), so this isn't a blocker but I think history lessons are way out of scope for an EIP.  The motivation section should be brief in explaining the value this standard adds to the ecosystem and leave it up to the user or the discussions-to for additional argumentation/context.

----

Test Cases section should be above the Security Considerations section.

----

Editor note: I just wanted to say how I appreciate your Test Cases section.  Most EIPs don't include any test cases, or if they do they are ill defined.  The test cases here are quite above the bar and I'm very happy to see them!

**Comment by gcolvin** [comment]
I'll take care of the issues noted by @lightclient.  Thanks.

I'll rework the dependencies.  The other EIPs are only needed to provide the degree of the completeness and safety that I was asked for last year, so I think I can just refer to them in the safety section rather than create a dependency.  This EIP could go in without them.

 @MicahZoltu, I think the references to Turing's original work need to stay.  Turing's document is a specification (for an actual computer) published in 1946 by the National Physical Laboratory of the United Kingdom -- best practices for real and virtual machines are long-established.  Turing's design remains a best practice, and a big part of my motivation is simply to bring the EVM into line with best practices.   So his specification directly informs this one, as I state in the both the Motivation and the Rationale.

**Comment by gcolvin** [comment]
As for having a History section at all.  In general, I agree.  But it became painfully clear to me last year that some critics of this proposal were not aware of the very long history of the well-established design that I am proposing here.

**Comment by gcolvin** [comment]
And all thanks go to @holiman for the Test Cases.

**Comment by holiman** [comment]
> And all thanks go to @holiman for the Test Cases.

Speaking of which. As of these new changes, the testcases are moot -- I think they should no longer be in the EIP. Alternatively, modified to the new semantics.

**Comment by gcolvin** [comment]
I've tried to modify them to account for the changes, @holiman.  We can deal with my mistakes during review.  As I recall it these got generated by the test code itself, so really getting them right means writing the test code.  Eventually I need to do that, but not unless there is consensus that we want to do this.

**Comment by gcolvin** [comment]
@lightclient @MicahZoltu   I've merged a new draft that tries to respond to your points.

File: EIPS/eip-2315.md
Comment (gumb0) [inline EIPS/eip-2315.md:32]:
```suggestion
Taken together with other required proposals they provide an _efficient, static, and safe_ control-flow facility.
```

**Comment by MicahZoltu** [comment]
> As for having a History section at all. In general, I agree. But it became painfully clear to me last year that some critics of this proposal were not aware of the very long history of the well-established design that I am proposing here.

The information is useful, just not in the EIP itself.  I recommend including it in the `discussions-to`.  Ethereum Magicians allows you to link directly to individual comments, so if anyone needs a history lesson you can just link them directly to the comment with the history lesson.

**Comment by gcolvin** [comment]
@MicahZoltu 
> The information is useful, just not in the EIP itself.

There is a tension here between an EIP being pure specification, and an EIP providing the information needed to evaluate and implement it.  As author - and my coauthor Brooklyn concurs - I really do insist that this section remain, at least through the Review period.  Note that we specifically call out the design we have chosen as Turing's design.  Without some context the reviewer may not understand.  So we have included a brief history and two stable links to the original source that we quote.

**Comment by axic** [comment]
> 4200 (Draft)

Filed #4716 to solve this.

**Comment by MicahZoltu** [comment]
> There is a tension here between an EIP being pure specification, and an EIP providing the information needed to evaluate and implement it.

You are correct on where the tension lies.  I am of the opinion that there is great value in having a document that contains useful information, history, links, discussion, etc. and one of those links would be the EIP.  In many cases, the EIP should *not* be the primary source of information on a particular topic.

I personally liked what Tim did for EIP-1559 which was to create this document that he maintained: https://hackmd.io/@timbeiko/1559-resources.  During the discussions around EIP-1559, I would link people to that document rather than the EIP itself, and if someone wanted they were able to click through to the EIP from there.  This gave Tim unconstrained decision making power over what was included in the document that was passed around to everyone and linked everywhere, while the EIP was able to remain pure throughout its lifecycle.

File: EIPS/eip-2315.md
Comment (lightclient) [inline EIPS/eip-2315.md:386]: This section is not valid per EIP-1 and should be removed.

**Review by lightclient — CHANGES_REQUESTED** [review]

File: EIPS/eip-2315.md
Comment (gcolvin) [reply EIPS/eip-2315.md:386]: Gone.

**Comment by gcolvin** [comment]
> I personally liked what Tim did for EIP-1559 ...

I did too.  This is a different situation.  There isn't that much extra material here, part of it is normative, and removing the rest would only make the Motivation appear weak and the Rationale appear arbitrary.

Recall that the previous version of this proposal was met with last-minute objections, semantic disputes, design disagreements, and demands for justifications that could not be answered directly from the EIP, or based on my authority as an expert.  I don't want to happen again.  So I cite an unimpeachable authority and try to make only statements that I can justify from my personal experience.  I can't count on people reading any document besides this EIP, so I need the relevant material here.

So if you cannot in good conscience move this EIP to Review, @MicahZoltu, I'd ask that you not block it and allow another editor to merge this PR, so these issues can be resolved by a larger consensus on the way to Final status.

**Pull request merged**
